zephyr/src/supp_api.c

```diff
@@ -0,0 +1,289 @@
+/**
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/logging/log.h>
+
+#include "includes.h"
+#include "common.h"
+#include "common/defs.h"
+#include "wpa_supplicant/config.h"
+#include "wpa_supplicant_i.h"
+#include "driver_i.h"
+
+#include "zephyr_fmac_main.h"
+#include "supp_main.h"
+#include "supp_api.h"
+
+int cli_main(int, const char **);
+extern struct wpa_global *global;
+
+enum requested_ops {
+	CONNECT = 0,
+	DISCONNECT
+};
+
+struct wpa_supp_api_ctrl {
+	const struct device *dev;
+	enum requested_ops requested_op;
+	int connection_timeout;
+};
+
+struct wpa_supp_api_ctrl wpa_supp_api_ctrl;
+
+static int send_wpa_supplicant_dummy_event(void)
+{
+	struct wpa_supplicant_event_msg msg = { 0 };
+
+	msg.ignore_msg = true;
+
+	return send_wpa_supplicant_event(&msg);
+}
+
+
+static inline struct wpa_supplicant * get_wpa_s_handle(const struct device *dev)
+{
+	struct wpa_supplicant *wpa_s = wpa_supplicant_get_iface(global, dev->name);
+	if (!wpa_s) {
+		wpa_printf(MSG_ERROR,
+		"%s: Unable to get wpa_s handle for %s\n", __func__, dev->name);
+		return NULL;
+	}
+
+	return wpa_s;
+}
+
+static void supp_shell_connect_status(void *wpa_supp_api_ctrl)
+{
+	int i = 0, status = 0;
+	struct wpa_supplicant *wpa_s;
+	struct wpa_supp_api_ctrl *ctrl = ctrl;
+
+	wpa_s = get_wpa_s_handle(ctrl->dev);
+	if (!wpa_s) {
+		status = 1;
+		goto out;
+	}
+
+	if (ctrl->requested_op == CONNECT) {
+		/* TODO: use a timer for connection timeout*/
+		while (wpa_s->wpa_state != WPA_COMPLETED &&
+			   (ctrl->connection_timeout == SYS_FOREVER_MS ||
+			    i++ < ctrl->connection_timeout))
+		{
+			k_yield();
```

Comment:
I don't think this yield is needed given that the thread is put to sleep just after.